pkg/apiserver/topo/topo.go

```diff
@@ -0,0 +1,251 @@
+		// handle error!
+		return nil
+	}
+
+	return &Service{etcd: cli, config: config}
+}
+
+func (s *Service) Register(r *gin.RouterGroup) {
+	endpoint := r.Group("/topo")
+	endpoint.GET("/", s.topologyHandler)
+}
+
+// @Summary Dashboard info
+// @Description Get information about the dashboard service.
+// @Produce json
+// @Success 200 {object} Topology
+// @Router /topo [get]
+func (s *Service) topologyHandler(c *gin.Context) {
+
+	dataMap := sync.Map{}
+	errMap := sync.Map{}
+	wg := sync.WaitGroup{}
+
+	namespace := "/topo/"
+	etcdPoints := []string{"alertmanager", "tidb", "grafana"}
+
+	for _, key := range etcdPoints {
+		wg.Add(1)
+		go func(k string) {
+			defer wg.Done()
+			v, err := s.etcdLoad(namespace + k)
+			if err != nil {
+				errMap.Store(k, err)
+				return
+			}
+			if v != "" {
+				dataMap.Store(k, v)
+			}
+		}(key)
+	}
+
+	// using pd http api to load pd peers and tikv peers
+	var pdAddresses []string
+	var kvAddresses []string
+
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		var err error
+		pdAddresses, err = s.pdLoad()
+		if err != nil {
+			errMap.Store("pd", err)
+		}
+	}()
+
+	go func() {
+		defer wg.Done()
+		var err error
+		kvAddresses, err = s.pdLoad()
+		if err != nil {
+			errMap.Store("tikv", err)
+		}
+	}()
+
+	wg.Wait()
+
+	for _, v := range etcdPoints {
+		e, exists := errMap.Load(v)
+		if !exists {
+			continue
+		}
+		err, ok := e.(error)
+		if !ok {
+			continue
+		}
+		c.JSON(http.StatusInternalServerError, err.Error())
+		return
+	}
+
+
+	c.JSON(http.StatusOK, Topology{
+		Grafana:      convert(dataMap.Load("grafana")),
+		Alertmanager: convert(dataMap.Load("alertmanager")),
+		TiKV:         pdAddresses,
+		TiDB:         parseArray(convert(dataMap.Load("tidb"))),
+		PD:           kvAddresses,
+	})
+}
+
+func convert(value interface{}, ok bool) string {
+	if !ok {
+		return ""
+	}
+	v, ok := value.(string)
+	if !ok {
+		// maybe we should return err
+		return ""
+	}
+	return v
+}
+
+// parseArray will parse addresses like "address1, address2, ..., addressN"
+// to array [address1, address2, ..., addressN].
+// If input is "", it will return [""]
+func parseArray(input string) []string {
+	array := make([]string, 0)
+	for _, s := range strings.Split(input, ",") {
+		array = append(array, strings.Trim(s, " "))
+	}
+	return array
+}
+
+// etcdLoad load key like "/topo/tidb" from pd's embedded etcd.
+// If the key doesn't exists, it will just return "", nil.
+// Otherwise, it will return value, nil.
+func (s *Service) etcdLoad(key string) (string, error) {
+	resp, err := s.etcd.Get(context.TODO(), key)
+	if err != nil {
+		return "", err
+	}
+	if resp == nil {
+		return "", nil
+	}
+
+	if len(resp.Kvs) == 0 {
+		return "", nil
+	}
+	return string(resp.Kvs[0].Value), nil
+}
+
+func (s *Service) tikvLoad() ([]string, error) {
+	stores := struct {
+		Count  int
+		Stores []struct {
+			Store struct {
+				Address string
+			}
+		}
+	}{}
+
+	resp, err := http.Get(s.config.PDEndPoint + "/pd/api/v1/members")
```

Comment:
this can also be done using etcd client?

Reply:
Well, I'm afraid not.

Reply:
Then is it possible to use pd client to do this?